libs/gru/src/main/java/com/agorapulse/gru/minions/HttpMinion.java

```diff
@@ -66,14 +83,25 @@ public void doVerify(Client client, Squad squad, GruContext context) {
             }
         }
 
-        for (Map.Entry<String, Collection<String>> header : responseHeaders.entrySet()) {
-            Optional.ofNullable(header.getValue()).ifPresent(headers -> {
-                for (String value : headers) {
-                    if (!(client.getResponse().getHeaders(header.getKey()).contains(value))) {
-                        throw new AssertionError("Missing header " + header.getKey() + " with value " + value);
+        for (Map.Entry<String, Collection<Matcher<String>>> headerMatcher : responseHeaderMatchers.entrySet()) {
+            List<String> headerValues = client.getResponse().getHeaders(headerMatcher.getKey());
+            if (Objects.isNull(headerValues) || headerValues.isEmpty()) {
+                throw new AssertionError("Missing header " + headerMatcher.getKey());
+            }
+
+            headerMatcher.getValue()
```

Comment:
This is a bit messy. Can someone take a look at this?

Reply:
have you run into some concurrent issues that we need a ConcurrentHashMap and a lock object?

Reply:
yes, please. if we face these issue we'll have to find out some other solutions how to handle concurrency.

Reply:
Sorry, i was on vacation. Just updated

Reply:
no worries 👍